From ba97df17866475c8062a71af524a6edb8c730a61 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Tue, 25 Aug 2020 17:20:19 +0300
Subject: [PATCH] [SW-17562] spi: move to common

In preparation to Gaudi security, flash OTP access is needed by preboot.

No changes (except trivial includes), only file move.

Change-Id: I2b3ae1bc8fd453ad462d1ae9b2bb23a74184fc13
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/common/efuse.c b/common/efuse.c
index 3bece08..b7f79aa 100644
--- a/common/efuse.c
+++ b/common/efuse.c
@@ -12,7 +12,7 @@
 #include "common.h"
 #include "preboot.h"
 #include "uart.h"
-#include "ppboot/spi.h"
+#include <spi.h>
 #include "pll.h"
 
 /*
diff --git a/common/spi.c b/common/spi.c
new file mode 100644
index 0000000..e5062ff
--- /dev/null
+++ b/common/spi.c
@@ -0,0 +1,243 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <stddef.h>
+
+#include <psoc_gpio3_regs.h>
+#include <psoc_spi_regs.h>
+
+#include "preboot.h"
+#include "common.h"
+#include <spi.h>
+#include "uart.h"
+
+/* Registers */
+#define CTRLR0				0x0
+#define SSIENR				0x8
+#define SER				0x10
+#define BAUDR				0x14
+#define TXFTLR				0x18
+#define RXFTLR				0x1C
+#define TXFLR				0x20
+#define RXFLR				0x24
+#define IMR				0x2C
+#define DR0				0x60
+
+#define CMD_WRITE_STATUS		0x01
+#define CMD_READ			0x03
+#define CMD_WRITE_DISABLE		0x04
+#define CMD_READ_STATUS			0x05
+#define CMD_WRITE_ENABLE		0x06
+#define frame_size_bits(b)		((b) << 16)
+
+static unsigned int spi_regs;
+/*
+ * SSI device Chip Select (CS0) is combined with GPIO53
+ * Using this we can really do SSI bus claim/release
+ * GPIO53 is set through mmPSOC_GPIO3 where each
+ * mmPSOC_GPIO serve 16 GPIOs
+ * Note: CS0 is active low
+ */
+__weak void spi_ssi_cs0_get(unsigned int *mask,
+					   unsigned int *data_offset,
+					   unsigned int *out_en_offset)
+{
+	*mask = BIT(53 % 16);
+	*data_offset = mmPSOC_GPIO3_DATAOUT;
+	*out_en_offset = mmPSOC_GPIO3_OUTENSET;
+}
+
+static void spi_ssi_cs0_init(void)
+{
+	unsigned int reg, mask, data, out_en;
+
+	spi_ssi_cs0_get(&mask, &data, &out_en);
+	reg  = regmap_readl(data);
+	reg  &= ~mask;
+	regmap_writel(reg, data);
+
+	reg = regmap_readl(out_en);
+	reg |=  mask;
+	regmap_writel(reg, out_en);
+}
+
+static void spi_claim_bus(void)
+{
+	regmap_writel(1, get_cs_pad_sel());
+}
+
+static void spi_release_bus(void)
+{
+	regmap_writel(0, get_cs_pad_sel());
+}
+
+static inline void spi_dev_update_frame_size(unsigned int len)
+{
+	regmap_writel(0, spi_regs + SSIENR);
+	regmap_writel(frame_size_bits((len * 8) - 1), spi_regs + CTRLR0);
+	regmap_writel(1, spi_regs + SSIENR);
+}
+
+static inline int spi_fifo_is_empty(unsigned int level_reg)
+{
+	return !regmap_readl(level_reg);
+}
+
+static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
+{
+	unsigned int tmp;
+
+	while (!spi_fifo_is_empty(spi_regs + TXFLR))
+		;
+	regmap_writel(cmd, spi_regs + DR0);
+
+	while (spi_fifo_is_empty(spi_regs + RXFLR))
+		;
+	tmp = regmap_readl(spi_regs + DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+static void spi_special_cmd(unsigned int cmd, unsigned char *val)
+{
+	unsigned char tmp;
+
+	spi_dev_update_frame_size(sizeof(tmp));
+
+	while (!spi_fifo_is_empty(spi_regs + TXFLR))
+		;
+	regmap_writeb(cmd, spi_regs + DR0);
+
+	while (spi_fifo_is_empty(spi_regs + RXFLR))
+		;
+	tmp = regmap_readb(spi_regs + DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+/* Hold chip select (CS) during transaction */
+static void spi_char_cmd_csel(unsigned char cmd, unsigned char *val)
+{
+	spi_claim_bus();
+
+	spi_special_cmd(cmd, NULL);
+	if (val)
+		spi_special_cmd(0, val);
+
+	spi_release_bus();
+}
+
+static void spi_special_cmd_csel(unsigned int cmd, unsigned char *val,
+								size_t size)
+{
+	int i;
+
+	spi_claim_bus();
+
+	spi_dev_update_frame_size(sizeof(*val));
+
+	while (!spi_fifo_is_empty(spi_regs + TXFLR))
+		;
+	regmap_writeb(cmd, spi_regs + DR0);
+	for (i = 0; i < size; i++) {
+		while (!spi_fifo_is_empty(spi_regs + TXFLR))
+			;
+		regmap_writeb(val[i], spi_regs + DR0);
+	}
+
+	while (spi_fifo_is_empty(spi_regs + RXFLR))
+		;
+	regmap_readb(spi_regs + DR0);
+	for (i = 0; i < size; i++) {
+		while (spi_fifo_is_empty(spi_regs + RXFLR))
+			;
+		regmap_readb(spi_regs + DR0);
+	}
+
+	spi_release_bus();
+}
+
+static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
+{
+	unsigned char sr;
+
+	spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
+	spi_special_cmd_csel(cmd, buf, size);
+	do
+		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	while (sr & READ_STATUS_WIP);
+	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
+}
+
+static unsigned char spi_prot(void)
+{
+	unsigned char sr;
+
+	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	sr |= PROT_MASK;
+	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr));
+
+	return sr;
+}
+
+static void spi_read(void *dst, unsigned int offset, unsigned int size)
+{
+	unsigned int cmd = (CMD_READ << 24) | offset;
+	int i;
+
+	spi_claim_bus();
+	spi_dev_update_frame_size(sizeof(cmd));
+	spi_dev_dword_cmd(cmd, NULL);
+	for (i = 0; i < size; i += sizeof(unsigned int))
+		spi_dev_dword_cmd(0, dst + i);
+	spi_dev_dword_cmd(0, NULL);
+	spi_release_bus();
+}
+
+__weak void spi_base_get(unsigned int *regs)
+{
+	*regs = mmPSOC_SPI_CTRLR0;
+}
+
+static void spi_init(void)
+{
+	u32 baudr = hl_spi_baudr_get();
+
+	/* Validate that baudr is even */
+	baudr = (baudr + 1) & ~1;
+
+	spi_base_get(&spi_regs);
+	spi_ssi_cs0_init();
+	regmap_writel(0x00000000, spi_regs + SSIENR);
+	regmap_writel(baudr, spi_regs + BAUDR);
+	regmap_writel(0x00000000, spi_regs + RXFTLR);
+	regmap_writel(0x00000000, spi_regs + TXFTLR);
+	regmap_writel(0x0000003F, spi_regs + IMR);	/* Disable Interrupts */
+	regmap_writel(0x00000001, spi_regs + SER);
+	regmap_writel(0x00000001, spi_regs + SSIENR);
+}
+
+void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name)
+{
+	unsigned char sr;
+	static bool prot_flash = true;
+
+	spi_init();
+	spi_read(dst, offset, size);
+	preboot_puts_raw(name);
+	preboot_puts(" copying from flash is completed\n");
+	if (prot_flash) {
+		prot_flash = false;
+		sr = spi_prot();
+		if ((sr & READ_STATUS_BP0_3) && (sr & READ_STATUS_SRWD))
+			preboot_puts("Flash may be protected, depend on WP# signal\n");
+		else
+			preboot_puts("Flash is not protected\n");
+	}
+}
diff --git a/gaudi/common.c b/gaudi/common.c
index a6e5559..3697522 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -22,7 +22,7 @@
 #include <pll.h>
 #include <uart.h>
 #include <string.h>
-#include <ppboot/spi.h>
+#include <spi.h>
 #include <preboot/fit.h>
 #include <preboot/thermal.h>
 
diff --git a/gaudi/common.h b/gaudi/common.h
index 749f983..ffb0b91 100644
--- a/gaudi/common.h
+++ b/gaudi/common.h
@@ -40,7 +40,7 @@
 #define I2C_SLAVE_ADDRESS		0x4f
 
 #ifndef __ASSEMBLY__
-#include "ppboot/spi.h"
+#include <spi.h>
 
 /* Let this file be included in any order and override the PROT_MASK */
 #ifdef PROT_MASK
diff --git a/goya/common.c b/goya/common.c
index 3a18a80..d5ebe9b 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -18,7 +18,7 @@
 #include <pll.h>
 #include <uart.h>
 #include <string.h>
-#include <ppboot/spi.h>
+#include <spi.h>
 #include <preboot/fit.h>
 #include <preboot/thermal.h>
 #include <string.h>
diff --git a/goya2/common.c b/goya2/common.c
index 3a12922..7030d77 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -23,7 +23,7 @@
 #include <pll.h>
 #include <uart.h>
 #include <string.h>
-#include <ppboot/spi.h>
+#include <spi.h>
 #include <preboot/fit.h>
 
 #include "common.h"
diff --git a/include/spi.h b/include/spi.h
new file mode 100644
index 0000000..faa710d
--- /dev/null
+++ b/include/spi.h
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2017 HabanaLabs Ltd.
+ *   Author: Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#ifndef _HL_SPI_H_
+#define _HL_SPI_H_
+
+/* The Status Register (SR) in use bits definition */
+#define READ_STATUS_WIP			BIT(0)
+#define READ_STATUS_BP0			BIT(2)
+#define READ_STATUS_BP1			BIT(3)
+#define READ_STATUS_BP2			BIT(4)
+#define READ_STATUS_BP3			BIT(5)
+#define READ_STATUS_SRWD		BIT(7)
+#define READ_STATUS_BP0_3		(READ_STATUS_BP0 | READ_STATUS_BP1 | \
+					READ_STATUS_BP2 | READ_STATUS_BP3)
+#define ALL_PROT_MASK			(READ_STATUS_BP0_3 | READ_STATUS_SRWD)
+
+/* The PROT_MASK is meant to be overridden by device specific setting */
+#ifndef PROT_MASK
+#define PROT_MASK			ALL_PROT_MASK
+#endif
+
+void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name);
+u32 get_cs_pad_sel(void);
+
+#endif /* _HL_SPI_H_ */
diff --git a/ppboot/boot_desc.c b/ppboot/boot_desc.c
index 35db95d..0bfc9a6 100644
--- a/ppboot/boot_desc.c
+++ b/ppboot/boot_desc.c
@@ -11,7 +11,7 @@
 #include <common.h>
 
 #include "boot_desc.h"
-#include "spi.h"
+#include <spi.h>
 #include "uart.h"
 
 #define MAX_IMGS		5
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 8aa738c..8c8baff 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -14,7 +14,7 @@
 
 #include "common.h"
 #include "uart.h"
-#include "spi.h"
+#include <spi.h>
 #include "boot_desc.h"
 #include "pll.h"
 
diff --git a/ppboot/spi.c b/ppboot/spi.c
deleted file mode 100644
index 74459ce..0000000
--- a/ppboot/spi.c
+++ /dev/null
@@ -1,243 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <stddef.h>
-
-#include <psoc_gpio3_regs.h>
-#include <psoc_spi_regs.h>
-
-#include "preboot.h"
-#include "common.h"
-#include "spi.h"
-#include "uart.h"
-
-/* Registers */
-#define CTRLR0				0x0
-#define SSIENR				0x8
-#define SER				0x10
-#define BAUDR				0x14
-#define TXFTLR				0x18
-#define RXFTLR				0x1C
-#define TXFLR				0x20
-#define RXFLR				0x24
-#define IMR				0x2C
-#define DR0				0x60
-
-#define CMD_WRITE_STATUS		0x01
-#define CMD_READ			0x03
-#define CMD_WRITE_DISABLE		0x04
-#define CMD_READ_STATUS			0x05
-#define CMD_WRITE_ENABLE		0x06
-#define frame_size_bits(b)		((b) << 16)
-
-static unsigned int spi_regs;
-/*
- * SSI device Chip Select (CS0) is combined with GPIO53
- * Using this we can really do SSI bus claim/release
- * GPIO53 is set through mmPSOC_GPIO3 where each
- * mmPSOC_GPIO serve 16 GPIOs
- * Note: CS0 is active low
- */
-__weak void spi_ssi_cs0_get(unsigned int *mask,
-					   unsigned int *data_offset,
-					   unsigned int *out_en_offset)
-{
-	*mask = BIT(53 % 16);
-	*data_offset = mmPSOC_GPIO3_DATAOUT;
-	*out_en_offset = mmPSOC_GPIO3_OUTENSET;
-}
-
-static void spi_ssi_cs0_init(void)
-{
-	unsigned int reg, mask, data, out_en;
-
-	spi_ssi_cs0_get(&mask, &data, &out_en);
-	reg  = regmap_readl(data);
-	reg  &= ~mask;
-	regmap_writel(reg, data);
-
-	reg = regmap_readl(out_en);
-	reg |=  mask;
-	regmap_writel(reg, out_en);
-}
-
-static void spi_claim_bus(void)
-{
-	regmap_writel(1, get_cs_pad_sel());
-}
-
-static void spi_release_bus(void)
-{
-	regmap_writel(0, get_cs_pad_sel());
-}
-
-static inline void spi_dev_update_frame_size(unsigned int len)
-{
-	regmap_writel(0, spi_regs + SSIENR);
-	regmap_writel(frame_size_bits((len * 8) - 1), spi_regs + CTRLR0);
-	regmap_writel(1, spi_regs + SSIENR);
-}
-
-static inline int spi_fifo_is_empty(unsigned int level_reg)
-{
-	return !regmap_readl(level_reg);
-}
-
-static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
-{
-	unsigned int tmp;
-
-	while (!spi_fifo_is_empty(spi_regs + TXFLR))
-		;
-	regmap_writel(cmd, spi_regs + DR0);
-
-	while (spi_fifo_is_empty(spi_regs + RXFLR))
-		;
-	tmp = regmap_readl(spi_regs + DR0);
-
-	if (val)
-		*val = tmp;
-}
-
-static void spi_special_cmd(unsigned int cmd, unsigned char *val)
-{
-	unsigned char tmp;
-
-	spi_dev_update_frame_size(sizeof(tmp));
-
-	while (!spi_fifo_is_empty(spi_regs + TXFLR))
-		;
-	regmap_writeb(cmd, spi_regs + DR0);
-
-	while (spi_fifo_is_empty(spi_regs + RXFLR))
-		;
-	tmp = regmap_readb(spi_regs + DR0);
-
-	if (val)
-		*val = tmp;
-}
-
-/* Hold chip select (CS) during transaction */
-static void spi_char_cmd_csel(unsigned char cmd, unsigned char *val)
-{
-	spi_claim_bus();
-
-	spi_special_cmd(cmd, NULL);
-	if (val)
-		spi_special_cmd(0, val);
-
-	spi_release_bus();
-}
-
-static void spi_special_cmd_csel(unsigned int cmd, unsigned char *val,
-								size_t size)
-{
-	int i;
-
-	spi_claim_bus();
-
-	spi_dev_update_frame_size(sizeof(*val));
-
-	while (!spi_fifo_is_empty(spi_regs + TXFLR))
-		;
-	regmap_writeb(cmd, spi_regs + DR0);
-	for (i = 0; i < size; i++) {
-		while (!spi_fifo_is_empty(spi_regs + TXFLR))
-			;
-		regmap_writeb(val[i], spi_regs + DR0);
-	}
-
-	while (spi_fifo_is_empty(spi_regs + RXFLR))
-		;
-	regmap_readb(spi_regs + DR0);
-	for (i = 0; i < size; i++) {
-		while (spi_fifo_is_empty(spi_regs + RXFLR))
-			;
-		regmap_readb(spi_regs + DR0);
-	}
-
-	spi_release_bus();
-}
-
-static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
-{
-	unsigned char sr;
-
-	spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
-	spi_special_cmd_csel(cmd, buf, size);
-	do
-		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	while (sr & READ_STATUS_WIP);
-	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
-}
-
-static unsigned char spi_prot(void)
-{
-	unsigned char sr;
-
-	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	sr |= PROT_MASK;
-	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr));
-
-	return sr;
-}
-
-static void spi_read(void *dst, unsigned int offset, unsigned int size)
-{
-	unsigned int cmd = (CMD_READ << 24) | offset;
-	int i;
-
-	spi_claim_bus();
-	spi_dev_update_frame_size(sizeof(cmd));
-	spi_dev_dword_cmd(cmd, NULL);
-	for (i = 0; i < size; i += sizeof(unsigned int))
-		spi_dev_dword_cmd(0, dst + i);
-	spi_dev_dword_cmd(0, NULL);
-	spi_release_bus();
-}
-
-__weak void spi_base_get(unsigned int *regs)
-{
-	*regs = mmPSOC_SPI_CTRLR0;
-}
-
-static void spi_init(void)
-{
-	u32 baudr = hl_spi_baudr_get();
-
-	/* Validate that baudr is even */
-	baudr = (baudr + 1) & ~1;
-
-	spi_base_get(&spi_regs);
-	spi_ssi_cs0_init();
-	regmap_writel(0x00000000, spi_regs + SSIENR);
-	regmap_writel(baudr, spi_regs + BAUDR);
-	regmap_writel(0x00000000, spi_regs + RXFTLR);
-	regmap_writel(0x00000000, spi_regs + TXFTLR);
-	regmap_writel(0x0000003F, spi_regs + IMR);	/* Disable Interrupts */
-	regmap_writel(0x00000001, spi_regs + SER);
-	regmap_writel(0x00000001, spi_regs + SSIENR);
-}
-
-void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name)
-{
-	unsigned char sr;
-	static bool prot_flash = true;
-
-	spi_init();
-	spi_read(dst, offset, size);
-	preboot_puts_raw(name);
-	preboot_puts(" copying from flash is completed\n");
-	if (prot_flash) {
-		prot_flash = false;
-		sr = spi_prot();
-		if ((sr & READ_STATUS_BP0_3) && (sr & READ_STATUS_SRWD))
-			preboot_puts("Flash may be protected, depend on WP# signal\n");
-		else
-			preboot_puts("Flash is not protected\n");
-	}
-}
diff --git a/ppboot/spi.h b/ppboot/spi.h
deleted file mode 100644
index faa710d..0000000
--- a/ppboot/spi.h
+++ /dev/null
@@ -1,30 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_SPI_H_
-#define _HL_SPI_H_
-
-/* The Status Register (SR) in use bits definition */
-#define READ_STATUS_WIP			BIT(0)
-#define READ_STATUS_BP0			BIT(2)
-#define READ_STATUS_BP1			BIT(3)
-#define READ_STATUS_BP2			BIT(4)
-#define READ_STATUS_BP3			BIT(5)
-#define READ_STATUS_SRWD		BIT(7)
-#define READ_STATUS_BP0_3		(READ_STATUS_BP0 | READ_STATUS_BP1 | \
-					READ_STATUS_BP2 | READ_STATUS_BP3)
-#define ALL_PROT_MASK			(READ_STATUS_BP0_3 | READ_STATUS_SRWD)
-
-/* The PROT_MASK is meant to be overridden by device specific setting */
-#ifndef PROT_MASK
-#define PROT_MASK			ALL_PROT_MASK
-#endif
-
-void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name);
-u32 get_cs_pad_sel(void);
-
-#endif /* _HL_SPI_H_ */
